internal/utils/utils.go

```diff
@@ -0,0 +1,34 @@
+// Package utils contains utility functions that are used throughout the project.
+package utils
+
+import (
+	"errors"
+	"os"
+)
+
+// FileExists checks if a file exists at the given path.
+func FileExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return false, err
+	}
+	return !errors.Is(err, os.ErrNotExist), nil
+}
+
+// IsDirEmpty checks if the specified directory is empty.
+func IsDirEmpty(path string) (bool, error) {
```

Comment:
>> On the file existence, you can
>
> Seems like something is missing here

Yes, scratch that, I removed part of the sentence after seeing the caller side (and why you had to check for the error afterwards)

> That's not a strong argument by itself IMO. I think the Go standard library still lacks a lot of functionality (especially when comparing it to Python's standard library). I like to add simple functions which can be re-used throughout the code into a utils package (once there are many, I split them off into `fileutil`, `stringutil`, `executil`, etc.).

I still think there is a very high value in following the patterns the community in general is following. One article on the util one for instance is https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common.

Go is not Python and I don’t write my Go code as my Python code. In Python, I’m following the python pattern, snake case and so on. In Go, I’m following the Go convention and idioms. This allows:
- better onboarding on new Go members who wants to take parts of the projects
- being able to dig into dependencies, in particular, where we are not the author
- have a sense of familiarity and expections when looking at other’s code

Most of the time, without good justifications, diverging from those patterns shows that the person is trying to port other language patterns to it.

I think those as a separate, imported package, would make sense, but clearly, don’t name that `utils` because it will become a mix of unrelated things, which will cross domain boundaries. We could have a package named around `fileutils` to scope them, which would be an improvement.

If we go that route and don’t import anything outside, I suggest though that we have this per domain package with then a complete testsuite on them (this could be easily 100% covered).

> While the functions are only used in one package, we can make them private functions of that package, but I expect that we will want to use these in other packages at some point, and having them as private functions makes them hard to discover.

While I still think we should keep that as private functions at first as we don’t use them anywhere else right now, in large projects in particular, we shouldn’t try to prepare for a future that never comes. This is how we build technical debts, because we will always be wrong. Start small and split up.

This is why I think the argument strongly stand. As I said, you can keep a separate package even if this is not my preferred option, but then, I’m asking to have a better, domain specific, package name and a testsuite for each of them.

Reply:
> Go is not Python and I don’t write my Go code as my Python code. 

Neither do I. I didn't say I want to write Go like Python, I said that Go's standard library is lacking and I see the need to extend "standard" functionality in a some way which allows it to be used throughout the project.

> In Python, I’m following the python pattern, snake case and so on. In Go, I’m following the Go convention and idioms. This allows:
> 
>     better onboarding on new Go members who wants to take parts of the projects
>     being able to dig into dependencies, in particular, where we are not the author
>     have a sense of familiarity and expections when looking at other’s code

I agree, but on the one hand, I see a difference between the community's efforts to describe a good and a bad go style (like https://go.dev/wiki/CodeReviewComments) and the opinions of individuals, even if they are respected in the community. On the other hand I see these as guidelines from which you can deviate if you have good reasons to do so.

> we shouldn’t try to prepare for a future that never comes. This is how we build technical debts, because we will always be wrong.

Fully agreed when it comes to added complexity, especially adding abstractions "because it will be useful in the future". I don't see the separate package adding complexity, but I'm fine with moving the functions to the broker package for now and split them off once we want to use them somewhere else.

Reply:
>  I'm fine with moving the functions to the broker package for now and split them off once we want to use them somewhere else.

I'm using it in the `password_test` package as well, so I moved it to it's own package again, named it `fileutils` and added tests.